api/repositories/pod_repository.go

```diff
@@ -55,62 +42,71 @@ type RuntimeLogsMessage struct {
-		"korifi.cloudfoundry.org/version": message.AppRevision,
+		korifiv1alpha1.CFAppGUIDLabelKey: message.AppGUID,
+		korifiv1alpha1.VersionLabelKey:   message.AppRevision,
 	})
 	if err != nil {
 		return nil, fmt.Errorf("failed to build labelSelector: %w", err)
 	}
-	listOpts := client.ListOptions{Namespace: message.SpaceGUID, LabelSelector: labelSelector}
 
-	pods, err := r.listPods(ctx, authInfo, listOpts)
+	podList := corev1.PodList{}
+	err = userClient.List(ctx, &podList, &client.ListOptions{Namespace: message.SpaceGUID, LabelSelector: labelSelector})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to list pods: %w", apierrors.FromK8sError(err, PodResourceType))
 	}
 
-	appLogs := make([]LogRecord, 0, int64(len(pods))*message.Limit/2)
-
-	k8sClient, err := r.userClientFactory.BuildK8sClient(authInfo)
+	logClient, err := r.userClientFactory.BuildK8sClient(authInfo)
 	if err != nil {
 		return nil, fmt.Errorf("failed to build user client: %w", err)
 	}
 
-	for _, pod := range pods {
-		var logReadCloser io.ReadCloser
-		logReadCloser, err = k8sClient.CoreV1().Pods(message.SpaceGUID).GetLogs(pod.Name, &corev1.PodLogOptions{Timestamps: true, TailLines: &message.Limit}).Stream(ctx)
-		if err != nil {
-			// untested
-			logger.Info("failed to fetch logs", "pod", pod.Name, "reason", err)
-			continue
-		}
+	logRecords := iter.Map(iter.Lift(podList.Items), func(pod corev1.Pod) iter.Iterator[LogRecord] {
+		return r.getPodLogs(ctx, logClient, logger, pod, message.Limit)
+	}).Collect()
 
-		r := bufio.NewReader(logReadCloser)
-		for {
-			var line []byte
-			line, err = r.ReadBytes('\n')
-			if err != nil {
-				if err == io.EOF {
-					break
-				} else {
-					_ = logReadCloser.Close()
-					return nil, fmt.Errorf("failed to parse pod logs: %w", err)
-				}
-			}
+	return iter.Chain(logRecords...).Collect(), nil
+}
+
+func (r *PodRepo) getPodLogs(ctx context.Context, k8sClient k8sclient.Interface, logger logr.Logger, pod corev1.Pod, limit int64) iter.Iterator[LogRecord] {
+	logReadCloser, err := k8sClient.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{Timestamps: true, TailLines: &limit}).Stream(ctx)
+	if err != nil {
+		logger.Info("failed to fetch logs", "pod", pod.Name, "reason", err)
+		return iter.Lift([]LogRecord{})
+	}
 
-			logRecord := lineToAppLogRecord(line)
+	defer logReadCloser.Close()
 
-			appLogs = append(appLogs, logRecord)
-		}
+	logLines := readLines(logReadCloser, logger.WithValues("pod", pod.Name))
+	return iter.Map(iter.Lift(logLines), logLineToLogRecord)
+}
 
-		_ = logReadCloser.Close()
+func readLines(r io.Reader, logger logr.Logger) []string {
+	lines := []string{}
+
+	var err error
+	var line []byte
+	bufReader := bufio.NewReader(r)
+	for {
+		line, err = bufReader.ReadBytes('\n')
+		lines = append(lines, string(line))
+		if err != nil {
+			if err != io.EOF {
+				logger.Info("failed to parse pod logs", "err", err)
+			}
+			break
+		}
```

Comment:
Using go-functional v2, see `it.Lines` - you can iterate over lines from a reader.

Reply:
```go
lines, errs := itx.Lines(r).Unzip()
if len(errs.Collect()) > 0 {
  panic("Ahhh")
}
lines.Collect()
```

Reply:
Or if you know reading cannot fail such as with reading from a string buffer:

```go
lines := itx.LinesString(r).Left().Collect()
```

Reply:
We wanted to get as much logs as we can regardless read errors. go-functional v1 LinesString did not look it supported that (i.e. it just failed) so we kept that code. When switching to v2 we would make use of `itx.LinesString` for sure